[14.0][ADD][WIP] purchase_vendor_bill_product_breakdown

ivs-cetmix opened the pull request.

ivs-cetmix committed:
[14.0][ADD] purchase_vendor_bill_product_breakdown

[FIX] Changed purchase order test

[FIX] Changed showing button 'Create Bill'. Changed button behavior on default.
Bug fixes.

[FIX] Changed invoice count compute. Tests fixed.

[FIX] Add new tests.

[FIX] Changed functions, field name. Changed purchase order line component view.

[FIX] Reformat code.

[FIX] Bug fix.

[FIX] Test chanted.

[FIX] Changed function _prepare_component_account_move_line.

[FIX] Changed Tests.

francesco-ooops reviewed (changes requested):
* In the **Purchase** tab of the Product there will be a button after the **Delivery Lead Time** in the Vendor lines

@isserver1 I can't see the button

francesco-ooops commented on purchase_vendor_bill_product_breakdown/readme/USAGE.rst:
```diff
@@ -0,0 +1,8 @@
+* Enable **Vendor Bill Breakdown** option in the **Purchase** section of the **Sales & Purchases** tab in Partner
+* In the **Purchase** tab of the Product there will be a button after the **Delivery Lead Time** in the Vendor lines
```
* In the **Purchase** tab of the Product there will be a button at the end of the Vendor lines (the button is displayed only for Vendors with "Vendor Bill Breakdown" enabled)

francesco-ooops commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo_component.py:
```diff
@@ -0,0 +1,35 @@
+    product_uom_qty = fields.Float(string="Qty per Unit", default=1, required=True)
+    product_uom_id = fields.Many2one(
+        "uom.uom",
+        string="Unit od Measure",
```
typo

francesco-ooops commented:
@geomer198 can you mark the PR as draft until ready for review?

geomer198 commented:
Why codecov/project check is failed?

geomer198 committed:
[FIX] fields parameter fix.

geomer198 committed:
[FIX] Model product.supplierinfo : added fields product_bill_components, product_variant_ids
product_bill_components - related with product_tmpl_id.bill_components
product_variant_ids - product_tmpl_id.product_variatn_ids

Model product.supplierinfo.component : added supplier and set a price by choised supplier.
Model product.template : added field bill_component.
Model purchase.order.line : creating components with supplier
Model purchase.order.line.component : Added supplier and price by supplier
Model res.partner : Changed bill_components field string

geomer198 committed:
[FIX] Tests refacrot.

geomer198 committed:
[FIX] Added test for func onchange_price().

geomer198 committed:
[FIX] Change product.supplierinfo.* models.

geomer198 committed:
[FIX] Changes:
* Model product.supplierinfo : changed action_open_component_view func.
* Model product.supplierinfo.component :
- added invalid_component_ids field
- deleted trash fields and methods
- changed _compute_current_price, onchange_component_id funcs.
* Model purchase.order.line : changed _update_purchase_order_line_components, action_open_component_view funcs.
* Model purchase.order.line.component :
    - added partner_id field
    - deleted trash fields and methods.
    - changed _compute_price_unit, onchange_component_id funcs.
Without changes tests

geomer198 committed:
[FIX] Changes:
* Fixed tests
* Model product.supplierinfo : added ‘price’ field
* Model product.supplierinfo.component : added ‘price_total’ field and changed onchange_component_id func.
 * Model purchase.order.line.component : changed price_unit field and onchange_component_id func.
* product.supplierinfo view : added price_total field

geomer198 committed:
[FIX] Changes:
* Model product.supplierinfo.component : changed onchange_component_id func
* TestProductSupplierinfoComponents : changed tests

geomer198 committed:
[FIX] TestComponentFlow : changed tests

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/account_move.py:
```diff
@@ -0,0 +1,22 @@
+    @api.model_create_multi
+    def create(self, vals_list):
+        # Handle component invoice line values
+        for key, vals in enumerate(vals_list, start=0):
```
What is the reason for using “enumerate” here?

geomer198 replied:
To change behavior on the first iteration.

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/account_move.py:
```diff
@@ -0,0 +1,22 @@
+class AccountMoveLine(models.Model):
+    _inherit = "account.move.line"
+
+    qty_components = fields.Float(string="Component Qty", default=0.0)
```
Field name should be “component_qty”. 
“String” key is not needed

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,46 @@
+    @api.depends("component_ids", "bill_components")
+    def _compute_price(self):
+        for rec in self.filtered(
+            lambda item: len(item.component_ids) > 0 and item.bill_components
```
Non need for len() here

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,46 @@
+    _inherit = "product.supplierinfo"
+
+    product_bill_components = fields.Boolean(related="product_tmpl_id.bill_components")
+    bill_components = fields.Boolean(related="name.bill_components")
```
“partner_bill_components” looks more explicit

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,46 @@
+        compute="_compute_product_variant_ids",
+        store=True,
+    )
+    price = fields.Float(compute="_compute_price", store=True, readonly=True)
```
You are using mapped “price_total” here. So field name should be “price_total” as well.
Precision should be added explicitly here to match system rounding settings 
Why do we need “read only” here? It is a computed field

geomer198 replied:
Maybe use store=True for the price field is not bad idea?

ivs-cetmix replied:
Yup

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,46 @@
+    component_ids = fields.One2many(
+        comodel_name="product.supplierinfo.component", inverse_name="supplierinfo_id"
+    )
+    product_variant_ids = fields.Many2many(
```
Why not to use related field here?

geomer198 replied:
If I try to use 'related' arg then getting error: "Type of related field product.supplierinfo.product_variant_ids is inconsistent with product.template.product_variant_ids"

ivs-cetmix replied:
This is because "product_variant_ids" is a o2m field not a m2m one. Why are you trying to use m2m field here?

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,46 @@
+            rec.product_variant_ids = rec.product_tmpl_id.product_variant_ids
+
+    def action_open_component_view(self):
+        """Open view with product components"""
```
Open product components view

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo_component.py:
```diff
@@ -0,0 +1,80 @@
+        comodel_name="res.partner", related="supplierinfo_id.name"
+    )
+    invalid_component_ids = fields.Many2many(
+        comodel_name="product.product",
```
Would be nice to have a “help” key with explanation of what is this field used for

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo_component.py:
```diff
@@ -0,0 +1,80 @@
+        compute="_compute_current_price",
+        store=True,
+        string="Current price",
+        readonly=True,
```
Do we need “read only” here?

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo_component.py:
```diff
@@ -0,0 +1,80 @@
+            supplier_id = rec.component_id.seller_ids.filtered(
+                lambda s: s.name == rec.supplierinfo_id.name
+            )
+            rec.write(
```
Use „update“ instead of „write“

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/product_supplierinfo_component.py:
```diff
@@ -0,0 +1,80 @@
+        """Set default value at component onchange"""
+        parent_component = self.supplierinfo_id.product_variant_ids
+        components = self.supplierinfo_id.component_ids.mapped("component_id")
+        self.write(
```
Never use „write“ in onchange. It might lead to cache issues. Use „update“ instead

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/purchase_order_line.py:
```diff
@@ -0,0 +1,185 @@
+        )
+
+    def _has_components(self):
+        return len(self.component_ids) > 0 and self.bill_components
```
No need for len() here

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/purchase_order_line.py:
```diff
@@ -0,0 +1,185 @@
+                line.qty_invoiced = line.last_qty_invoiced
+
+    @api.model
+    def _compute_invoice_qty(self, qty_components, invoice_count):
```
Are you computing quantity of invoices or quantity of components invoced?
if the latter then „_compute_invoiced_qty“ will be more logical

geomer198 replied:
Resolved

ivs-cetmix commented on purchase_vendor_bill_product_breakdown/models/purchase_order_line.py:
```diff
@@ -0,0 +1,185 @@
+        suppliers = self.product_id.seller_ids.filtered(
+            lambda seller: seller.name.id == partner_id
+        )
+        return suppliers[0] if len(suppliers) > 0 else False
```
No need for “Len” here

geomer198 replied:
Resolved

geomer198 committed:
[FIX] Model account.move.line : renamed 'qty_component' field to 'component_qty'.

geomer198 committed:
[FIX] Model product.supplierinfo : rename bill_components field to partner_bill_components

geomer198 committed:
[FIX] Model product.supplierinfo.component :
- Added hrlp param to the invalid_compoent_ids field.
- Changed _compute_current_price, onchange_component_id funcs.

geomer198 committed:
[FIX] Model purchase.order.line :
- Changed get_supplier, _has_components, _compute_last_qty_invoiced funcs.
- Renamed _compute_invoice_qty func to _compute_invoiced_qty.

ivs-cetmix reviewed (commented):
Please solve the pending issues. After that open a new PR and I will close this one

geomer198 committed:
[FIX] Model product.supplierinfo : Changed product_variant_ids field type.
